### ISSUE TITLE
musl-based DNS resolution will break on v1.25.0 in certain configurations

### ISSUE THREAD

**Issue by dghubble** [kubernetes/kubernetes#112135]
### What happened?

This one involves systemd, DNS, musl, and Kubelet v1.25 😬 👷🏻 🔥 

Starting in v1.25.0, pods running on Kubelets on certain nodes with a FQDN hostname (more on that later) will not be able to resolve DNS queries if they're using musl (e.g. alpine:3.16.2, busybox, etc).

```
# OK
kubectl run debug -it --image docker.io/busybox:1.34.1 --rm=true -- wget www.google.com
kubectl run debug -it --image docker.io/busybox:1.34.1-uclibc --rm=true -- wget www.google.com
kubectl run debug -it --image docker.io/ubuntu:22.04 --rm=true -- /bin/sh -c "apt-get update; apt-get install -y wget & wget www.google.com"

# FAIL
kubectl run debug -it --image docker.io/busybox:1.34.1-musl --rm=true -- wget www.google.com
kubectl run debug -it --image docker.io/alpine:3.16.2 --rm=true -- wget www.google.com
```

Our story starts with systemd. Back in 2020, systemd [started populating](https://github.com/systemd/systemd/pull/17201/files) a "search ." in stub resolvers *if* it detected the hostname to be a FQDN (to solve a separate problem). Check for a static hostname with `hostnamectl`.

In https://github.com/kubernetes/kubernetes/pull/109441, a change was made to Kubelet to propagate that "search ." into pods (to solve a separate DNS problem). That change landed in `v1.25.0-alpha.1` and appears in `1.25.0`.

Unfortunatey, musl based DNS resolvers struggle with this "search ." EDIT: details in https://github.com/coreos/fedora-coreos-tracker/issues/1287#issuecomment-1232763947

### What did you expect to happen?

musl-based DNS resolvers (depite their other quirks) should continue to function 

### How can we reproduce it (as minimally and precisely as possible)?

To see if you have a suitable repro node, check your `/etc/resolv.conf` (likely a symlink to `/run/systemd/resolve/stub-resolv.conf`.

```
nameserver 127.0.0.53
options edns0 trust-ad
search .                        <--- this is populated by systemd if your hostname is a FQDN
```

Next, start a container (any CNI) that makes some use of musl resolvers. Confirm that starting in Kubelet v1.25.0, the container `/etc/resolv.conf` now includes the "search ." if it was present on the host.

```
kubectl run debug -it --image docker.io/alpine:3.16.2 --rm=true -- /bin/ash
```

```
search default.svc.cluster.local svc.cluster.local cluster.local .        <--- see this dot at the end
nameserver 10.3.0.10
options ndots:5
```

Try using musl-based DNS resolution.

```
apk update                    # FAILS
wget www.google.com  # FAILS
```

Note that `nslookup` **uses its own DNS resolver**, its not a reproducer. Note also that pod-to-pod traffic and outbound masquerade are fine too. This is not a masquerade problem.

```
nslookup www.google.com
Server:         10.3.0.10
Address:        10.3.0.10:53

Name:   www.google.com
Address: 2607:f8b0:4005:810::2004

Name:   www.google.com
Address: 142.250.189.164
```

Edit the `/etc/resolv.conf` in the container to remove the "." and your musl problems instantly disappear. Add the dot and they're back.

### Anything else we need to know?

I've been spelunking for a while.

* I can confirm this is not related to a specific CNI - flannel, Calico, and Cilium all reproduce the issue. Not networking related
* OpenShift is not affected. In 2021, RedHat folks noticed DNS "breaks" and reverted the systemd change on their OS https://github.com/openshift/okd-machine-os/pull/159

### Folks

cc @Miciah @vrutkovs

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
```

</details>

### Cloud provider

<details>
This isn't cloud provider specific. However, the most likely cases where folks use FQDN's are on-prem imo. So you're likely to see reports originating from there.
</details>

### OS version

<details>
```console
NAME="Fedora Linux"
VERSION="36.20220806.3.0 (CoreOS)"
```
</details>

### Install tools

<details>
This was found while preparing Typhoon's v1.25.0 release of Kubernetes v1.25.0
</details>

### Container runtime (CRI) and version (if applicable)

<details>
containerd 1.6.6
</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by dghubble** [comment]
Could either revert https://github.com/kubernetes/kubernetes/pull/109441 (which propagates "search ." into containers) or find a different solution to the empty "search " problem it was trying to solve (does OpenShift even need that fix if they already [strip "search ." from hosts](https://github.com/kubernetes/kubernetes/pull/109441)? cc @rphillips @vrutkovs @Miciah

A workaournd to problem is to configure hosts to strip "search ." from `/etc/resolv.conf` ala https://github.com/openshift/okd-machine-os/pull/159

**Comment by dghubble** [comment]
https://github.com/coreos/fedora-coreos-tracker/issues/1287#issuecomment-1232763947 has a great packet capture and this has been reported to musl https://www.openwall.com/lists/musl/2022/08/31/1

**Comment by aojea** [comment]
impressive investigation in the linked issue https://github.com/coreos/fedora-coreos-tracker/issues/1287#issuecomment-1233041103 , that seems to conclude that is a muslc bug, since it happens outside of Kubernetes too.

Just reverting the PR in Kubernetes doesn't looks right, it fixed a known bug ... and this is really a muslc bug .... 
maybe always removing the single dot `.` from `search` ? 

I think that for Kubernetes we should understand first what `search .` does and means and its implications, maybe some `libc` developer can help? @dcbw you know someone?

**Comment by dghubble** [comment]
Its discussed by the musl folks what `.` means. It does not sound trivial to implement on the musl side, its a new case they did not anticipate supporting. Unfortunately this is issue subtle, and folks using alpine or related ecosystem things will get bitten by this. I agree its a musl issue, but Kubelet now propagating `search .` triggers this. 

I think its worth noting that https://github.com/kubernetes/kubernetes/pull/109441 did not set out to propagate `search .`, it was merely trying to prevent an empty `search`. There may be better ways to achive that.

**Comment by thockin** [comment]
IIUC `search .` means "try the input name without any suffix".  Resolvers will do this ANYWAY, but it's implementation-defined whether they do it first or last or in parallel, so allowing `.` lets node owners be explicit (though, curiously, it seems at leat glibc may try the same name AGAIN at the end of the search list).

This is pretty clearly a bug in MUSL, but I wonder if, instead of https://github.com/kubernetes/kubernetes/pull/109441, we should have treated `search .` as `[]` rather than `["."]` or the errant `[""]` ?  It nullifies the control that `.` provides, it wouldn't tickle musl like this.

```
root@a53f7fc98ed6:/# cat /etc/resolv.conf 
search foo1 . foo2
nameserver 8.8.8.8

root@a53f7fc98ed6:/# dig +question +showsearch N | grep "^;N."
;N.foo1.				IN	A
;N.				IN	A
;N.foo2.				IN	A
;N.				IN	A
```

**Comment by aojea** [comment]
/cc @chrisohaver 

I foresee a new wave of bug reports caused because of this muslc bug , better to keep CoreDNS friends in the loop

**Comment by lucab** [comment]
The patch for this got merged for the upcoming 1.25.1 and 1.26
